File: Recursion/SwapNodesInPairs.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class LinkedList:
    def __init__(self):
        self.head = None
    
    def traverseList(self):
        while (self.head):
            print(self.head.val)
            self.head=self.head.next

class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        length = self.getLength(head)
        dummy = ListNode(0, head)
        prev = dummy
        curr = head

        for _ in range(length//2):
            next = curr.next
            curr.next = next.next
            next.next = prev.next
            prev.next = next
            prev = curr
            curr = curr.next

        return dummy.next

    def getLength(self ,head: ListNode)->int:
        length = 0
        while head:
            length+=1
            head=head.next
        return length

if __name__== "__main__":
    head = LinkedList()
    node1 = ListNode(1)
    node2= ListNode(2)
    node3 = ListNode(3)
    node4= ListNode(4)

    node1.next = node2
    node2.next = node3
    node3.next = node4

    head.head = node1
    Solution().swapPairs(head)
    head.traverseList()

File: Recursion/SwapNodesInPairs.cpp
// https://leetcode.com/explore/featured/card/recursion-i/250/principle-of-recursion/1681/
#include<bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        const int length = getLength(head);
        ListNode dummy(0 ,head);
        ListNode* prev = &dummy;
        ListNode* curr = head;

        for(int i=0;i<length/2;i++){
            ListNode* next = curr->next;
            curr->next = next->next;
            next->next = prev->next;
            prev->next = next;
            prev = curr;
            curr = curr->next;
        }

        return dummy.next;
    }

private:
    int getLength(ListNode* head){
        int length = 0;
        for(ListNode* curr= head; curr; curr = curr->next) ++length;
        return length;
    }
};

int main(){

    return 0;
}